orchestrator/relayer/src/fee_manager.rs

```diff
@@ -0,0 +1,219 @@
+use ethereum_gravity::{
+    utils::GasCost,
+};
+use ethers::prelude::*;
+use ethers::types::Address as EthAddress;
+use gravity_utils::ethereum::{format_eth_address};
+use gravity_utils::types::{ Erc20Token, };
+use std::collections::HashMap;
+use std::time::Duration;
+use std::time::Instant;
+use gravity_utils::types::config::RelayerMode;
+use reqwest::Error;
+
+const DEFAULT_TOKEN_PRICES_PATH: &str = "token_prices.json";
+const DEFAULT_TOKEN_ADDRESSES_PATH: &str = "token_addresses.json";
+const DEFAULT_RELAYER_API_URL:  &str = "https://cronos.org/gravity-testnet2/api/v1/oracle/quotes";
+
+pub struct FeeManager {
+    token_price_map: HashMap<String, String>,
+    token_api_url_map: HashMap<String, String>,
+    next_batch_send_time: HashMap<EthAddress, Instant>,
+    mode: RelayerMode,
+}
+
+#[derive(serde::Deserialize, Debug)]
+struct ApiResponse {
+    status: String,
+    result: ApiResult,
+}
+
+#[derive(serde::Deserialize, Debug)]
+struct ApiResult {
+    tokenPairDecimal: u32,
+    aggregatedPrice: f64,
+}
+
+impl FeeManager {
+    pub async fn new_fee_manager(mode: RelayerMode) -> Result<FeeManager, ()> {
+        let mut fm =  Self {
+            token_price_map: Default::default(),
+            token_api_url_map: Default::default(),
+            next_batch_send_time: HashMap::new(),
+            mode
+        };
+
+        let success = fm.init().await.unwrap();
+        if !success {
+            return Err(error!("Could not initialize fee manager"));
+        }
+        return Ok(fm);
+    }
+
+    async fn init(&mut self) -> Result<bool, ()> {
+        match self.mode {
+            RelayerMode::Api => {
+                return self.initWithApi().await;
+            }
+            RelayerMode::File => {
+                return self.initWithFile().await;
+            }
+            RelayerMode::AlwaysRelay => {
+                return Ok(true);
+            }
+            _ => {
+                return Ok(false);
+            }
+        }
+    }
+
+    async fn initWithApi(&mut self) -> Result<bool, ()> {
+        let token_addresses_path =
+            std::env::var("TOKEN_ADDRESSES_JSON").unwrap_or_else(|_| DEFAULT_TOKEN_ADDRESSES_PATH.to_owned());
+
+        let token_addresses_str = match tokio::fs::read_to_string(token_addresses_path).await {
+            Err(err) => {
+                return Err(error!("Error while fetching token pair addresses {}", err));
+            }
+            Ok(value) => value,
+        };
+
+        let token_addresses: HashMap<String, String> = match serde_json::from_str(&token_addresses_str)
+        {
+            Err(err) => {
+                return Err(error!("Error while parsing token pair addresses json configuration: {}", err));
+            }
+            Ok(token_addresses) => token_addresses,
+        };
```

Comment:
also can use `map_err`

Reply:
I've tried to use 

```
match serde_json::from_str(&token_addresses_str).map_err(|e| {
                error!(xxx);
                ()
            })?;
```

syntax here but does not seems to work. How do you use `map_err` here?

Reply:
nvm found it

Reply:
like this:
```
let token_address: HashMap<String, String> = serde_json::from_str(xxxxx).map_err(|e| .....)?;
```